components/codec/src/columnar/timestamp.rs

```diff
@@ -11,3 +11,301 @@
+
+const ENCODE_VERSION: u8 = 1;
+const NUM_BYTES_ENCODE_VERSION_LEN: usize = 1;
+
+const NUM_BYTES_TIMESTAMP_LEN: usize = 4;
+const NUM_BYTES_TIMESTAMP: usize = 8;
+
+/// END_MARKER is a special bit sequence used to indicate the end of the
+/// stream
+const END_MARKER: u64 = 0b1111_0000_0000_0000_0000_0000_0000_0000_0000;
+/// END_MARKER_LEN is the length, in bits, of END_MARKER
+const NUM_BITS_END_MARKER_LEN: usize = 36;
+
+// Encode the timestamp using the delta-of-delta algorithm
+struct TimestampEncoder {}
+
+struct TimestampDecoder {}
+
+enum DeltaOfDeltaEncodeMasks {
+    Zero(),
+    Normal(u8, u8, u8),
+    None(),
+}
+
+impl TimestampEncoder {
+    fn encode_delta_of_delta_masks(dod: i64) -> DeltaOfDeltaEncodeMasks {
+        // max support 2^31/3600/1000/24/2=12 day
+        if dod > std::i32::MAX as i64 {
+            debug!(
+                "dod reach max value, fallback to origin timestamp encode, dod:{:?}",
+                dod
+            );
+            return DeltaOfDeltaEncodeMasks::None();
+        }
+        let dod = dod as i32;
+
+        // store the delta of delta using variable length encoding
+        match dod {
+            0 => DeltaOfDeltaEncodeMasks::Zero(),
+            -63..=64 => DeltaOfDeltaEncodeMasks::Normal(0b10, 2, 7),
+            -255..=256 => DeltaOfDeltaEncodeMasks::Normal(0b110, 3, 9),
+            -2047..=2048 => DeltaOfDeltaEncodeMasks::Normal(0b1110, 4, 12),
+            _ => DeltaOfDeltaEncodeMasks::Normal(0b1111, 4, 32),
+        }
+    }
+
+    pub fn encode<B, I>(buf: &mut B, values: I) -> Result<()>
+    where
+        B: BufMut,
+        I: Iterator<Item = i64>,
+    {
+        let (lower, higher) = values.size_hint();
+        let num = lower.max(higher.unwrap_or_default());
+        ensure!(
+            num >= 1,
+            NotEnoughDatums {
+                expect: 1usize,
+                found: 0usize
+            }
+        );
+
+        let mut writer = BufferedWriter::with_capacity(0);
+
+        // write version
+        writer.write_bits(
+            ENCODE_VERSION as u64,
+            (8 * NUM_BYTES_ENCODE_VERSION_LEN) as u32,
+        );
+
+        // variables used for none encode timestamp
+        let mut need_fallback = false;
+        let mut dod_encode_ts_len = 1;
+
+        // variables used to encode dod
+        let mut first = true;
+        let mut last_ts: i64 = 0;
+        let mut last_delta: i64 = 0;
+
+        for v in values {
+            if !need_fallback {
+                if first {
+                    // store the first timestamp exactly
+                    writer.write_bits(v as u64, 64);
+                    first = false;
+                    last_ts = v;
+                    if num == 1 {
+                        writer.write_bit(Bit(0))
+                    } else {
+                        writer.write_bit(Bit(1))
+                    }
+                    continue;
+                }
+                let cur_delta = v - last_ts;
```

Comment:
Shall we consider the overflow?

Reply:
No need, `cur_delta` only overflow when the two timestamp is negative, I think we do not need to consider this situation.

Reply:
`delta` will over flow if `v - last_ts` > ` 9223372036854775807`,  which means the time difference is `922337203685477` seconds, It is impossible to appear in normal time series data.